routing/payment_lifecycle_test.go

```diff
@@ -502,33 +565,44 @@ func TestRouterPaymentStateMachine(t *testing.T) {
 				routerFailAttempt,
 				routerFailAttempt,
 
+				// shard 3
+				routeRelease,
+				routerRegisterAttempt,
+				sendToSwitchSuccess,
+
 				// The fourth shard succeed against all odds,
 				// making the overall payment succeed.
 				getPaymentResultSuccess,
 				routerSettleAttempt,
 				paymentSuccess,
 			},
-			routes: []*route.Route{shard, shard, shard, shard},
+			routes: []*route.Route{shard, shard, shard, rt},
 		},
 		{
 			// An MP payment scenario a shard fail with a terminal
 			// error, causing the router to stop attempting.
+			name: "MP terminal",
```

Comment:
When I ran this test multiple times, sometime I would get a fail. Curious if it ever happen to you? 
```
    // The first shard fail with a terminal error.
    getPaymentResultTerminalFailure,
    routerFailAttempt,
    routerFailPayment,
```
After this setup, the shardHandler will have a terminal state and it will shut down, causing the following steps failed,
```
    // Remaining 3 shards fail.
    getPaymentResultTempFailure,
    getPaymentResultTempFailure,
    getPaymentResultTempFailure,
    routerFailAttempt,
    routerFailAttempt,
    routerFailAttempt,
```

Reply:
Did this happen in this commit specifically? I think that this test case also hits the bug that is fixed in the following commits, which may be causing the failure. Found a few other races in these tests this morning, so will push a new version with fixes soon.

Reply:
Nope. This happens in another PR that I made, after refactoring the `shardHandler`. I think you got it fixed here. The problem was multiple shards were running, one of them got a terminal error causing it to shut down, depending on the timing, it may cause error in other shards.